toolbox/relnotes/main.go

```diff
@@ -0,0 +1,606 @@
+	preview       = flag.Bool("preview", false, "Report additional branch statistics (used for reporting outside of releases)")
+	releaseBucket = flag.String("release-bucket", "kubernetes-release", "Specify gs bucket to point to in generated notes (informational only)")
+	releaseTars   = flag.String("release-tars", "", "Directory of tars to sha256 sum for display")
+	repo          = flag.String("repo", "kubernetes", "Github repository")
+
+	// Global
+	branchHead = ""
+)
+
+func main() {
+	// Initialization
+	flag.Parse()
+	branchRange := flag.Arg(0)
+
+	log.Printf("Boolean flags: full: %v, htmlize-md: %v, preview: %v", *full, *htmlizeMD, *preview)
+	log.Printf("Input branch range: %s", branchRange)
+
+	// If branch isn't specified in flag, use current branch
+	if *branch == "" {
+		var err error
+		*branch, err = u.GetCurrentBranch()
+		if err != nil {
+			log.Printf("not a git repository: %s", err)
+			return
+		}
+	}
+
+	prFileName := "/tmp/release-notes-" + *branch + "prnotes"
+	if *mdFileName == "" {
+		*mdFileName = "/tmp/release-notes-" + *branch + ".md"
+	}
+
+	// If githubToken isn't specified in flag, use the GITHUB_TOKEN environment variable
+	if *githubToken == "" {
+		*githubToken = os.Getenv("GITHUB_TOKEN")
+	}
+	client := u.NewClient(*githubToken)
+
+	// Get release related commits on the release branch within release range
+	releaseCommits, startTag, releaseTag, err := getReleaseCommits(client, *owner, *repo, *branch, branchRange)
+	if err != nil {
+		log.Printf("failed to get release commits for %s: %s", branchRange, err)
+		return
+	}
+
+	// Parse release related PR ids from the release commits
+	commitPRs, err := parsePRFromCommit(releaseCommits)
+	if err != nil {
+		log.Printf("failed to parse release commits: %s", err)
+		return
+	}
+
+	// Get number-issue mapping for issues in the repository
+	issues, err := u.ListAllIssues(client, *owner, *repo)
+	if err != nil {
+		log.Printf("failed to list all issues from %s: %s", *repo, err)
+		return
+	}
+	issueMap := make(map[int]*github.Issue)
+	for _, i := range issues {
+		issueMap[*i.Number] = i
+	}
+
+	// Get release note PRs by examining release-note label on commit PRs
+	releasePRs := make([]int, 0)
+	for _, pr := range commitPRs {
+		if u.HasLabel(issueMap[pr], "release-note") {
+			releasePRs = append(releasePRs, pr)
+		}
+	}
+
+	// Generating release note...
+	log.Printf("Generating release notes...")
+
+	prFile, err := os.Create(prFileName)
+	if err != nil {
+		log.Printf("failed to create release note file %s: %s", prFileName, err)
+		return
+	}
+
+	// Bootstrap notes for minor (new branch) releases
+	if *full || u.IsVer(releaseTag, "dotzero") {
+		draftURL := u.GithubRawURL + *owner + "/features/master/" + *branch + "/release-notes-draft.md"
+		changelogURL := u.GithubRawURL + *owner + "/" + *repo + "/master/CHANGELOG.md"
+		minorRelease(prFile, releaseTag, draftURL, changelogURL)
+	} else {
+		patchRelease(prFile, startTag, releasePRs, issueMap)
+	}
+
+	prFile.Close()
```

Comment:
You need to use defer functions to handle something like close files, it's a golang trick. Take a look at [a example](https://github.com/istio/test-infra/blob/898e63af2a775b5038a4d5eb12d324ec4c0047bb/toolbox/pkg_check/main.go#L130)

Reply:
@yutongz The problem is that I need to close those files early, for

1. opening prFile again and copy the content into mdFile
2. running sed on mdFile to htmlize it
3. running pandoc on mdFile 

I will rearrange `main()` to be able to call defer functions.

Reply:
Done.